src/Constraints/ArraySameWithDates.php

```diff
@@ -0,0 +1,64 @@
+<?php
+declare(strict_types=1);
+
+namespace Eonx\TestUtils\Constraints;
+
+use DateTimeInterface;
+use PHPUnit\Framework\Constraint\Constraint;
+use PHPUnit\Framework\Constraint\IsIdentical;
+
+class ArraySameWithDates extends Constraint
+{
+    /**
+     * @var mixed[]
+     */
+    private $expected;
+
+    /**
+     * ArraySameWithDates constructor.
+     *
+     * @param mixed[] $expected
+     */
+    public function __construct(array $expected)
+    {
+        $this->expected = $expected;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function evaluate($other, string $description = '', bool $returnResult = false)
+    {
+        // If array can be compared without invoking constraint,
+        // that's the best path.
+        if ($this->expected === $other) {
+            return true;
+        }
+
+        $format = static function (&$value): void {
+            if (($value instanceof DateTimeInterface) === true) {
+                /**
+                 * @var \DateTimeInterface $value
+                 *
+                 * @see https://youtrack.jetbrains.com/issue/WI-37859 - typehint required until PhpStorm recognises ===
+                 */
+                $value = $value->format(DateTimeInterface::RFC3339);
+            }
+        };
+
+        \array_walk_recursive($this->expected, $format);
+        \array_walk_recursive($other, $format);
+
+        // Check that two arrays are identical.
+        $constraint = new IsIdentical($this->expected);
```

Comment:
is it OK to call constraints from within constraints like this? I could not see any such usage in phpunit's constraints.
I see they use `Comparators`, but they only `assertEquals` and not `assertSame`.

Reply:
Yep, seems to match the phpunit built in patterns.